### ISSUE TITLE
how to run a disk-based local node with TLS certs

### ISSUE THREAD

**Issue by glycerine** [cockroachdb/cockroach#4190]
I'm just trying to validate on a single local node, with disk storage and TLS certs.

Yet when I startup, cockroachdb complains about invalid cert.  How does one tell the db which ip address to use?

~~~
jaten@Jasons-MacBook-Pro:~/go/src/github.com/cockroachdb/cockroach$ git log|head
commit e83c0939857331471e913ba97a8b821b906161d9
Merge: 20e68e7 ef157d1
Author: Peter Mattis <peter@cockroachlabs.com>
Date:   Sat Feb 6 15:58:30 2016 -0500

    Merge pull request #4185 from petermattis/pmattis/bootstrap-fail
    
    Stop make if the .bootstrap target fails.

commit ef157d1b76e0ce8b81476b8b8d855adde7e0a1d7
jaten@Jasons-MacBook-Pro:~/go/src/github.com/cockroachdb/cockroach$ go build
jaten@Jasons-MacBook-Pro:~/go/src/github.com/cockroachdb/cockroach$ rm -rf certs
jaten@Jasons-MacBook-Pro:~/go/src/github.com/cockroachdb/cockroach$ cockroach cert create-ca
jaten@Jasons-MacBook-Pro:~/go/src/github.com/cockroachdb/cockroach$ cockroach cert create-node 127.0.0.1
jaten@Jasons-MacBook-Pro:~/go/src/github.com/cockroachdb/cockroach$ cockroach cert create-client root
jaten@Jasons-MacBook-Pro:~/go/src/github.com/cockroachdb/cockroach$ rm -rf /Users/jaten/cockroachdb.data
jaten@Jasons-MacBook-Pro:~/go/src/github.com/cockroachdb/cockroach$ cockroach start --stores=hdd=/Users/jaten/cockroachdb.data 
I0206 17:11:06.006020 49646 cli/start.go:104  [build]  @  (go1.5.1)
I0206 17:11:06.006447 49646 server/context.go:218  1 storage engine(s) specified
I0206 17:11:06.006467 49646 cli/start.go:125  starting cockroach node
I0206 17:11:06.012203 49646 storage/engine/rocksdb.go:106  opening rocksdb instance at "/Users/jaten/cockroachdb.data"
I0206 17:11:06.018714 49646 server/node.go:320  store store=0:0 ([hdd]=/Users/jaten/cockroachdb.data) not bootstrapped
I0206 17:11:06.024081 49646 storage/replica_command.go:1215  range 1: new leader lease replica {1 1 1} 1970-01-01 00:00:00 +0000 UTC +1454807467.023s
I0206 17:11:06.024793 49646 server/node.go:263  **** cluster b2a4b220-9503-419e-acc0-7c0c7f31d6f9 has been created
I0206 17:11:06.024834 49646 server/node.go:264  **** add additional nodes by specifying --join=[::]:26257
I0206 17:11:06.026313 49646 server/node.go:333  initialized store store=1:1 ([hdd]=/Users/jaten/cockroachdb.data): {Capacity:367416967168 Available:150502281216 RangeCount:0}
I0206 17:11:06.026548 49646 server/node.go:236  node ID 1 initialized
I0206 17:11:06.026989 49646 gossip/gossip.go:218  setting node descriptor node_id:1 address:<network_field:"tcp" address_field:"[::]:26257" > attrs:<> 
I0206 17:11:06.031290 49646 storage/stores.go:277  read 0 node addresses from persistent storage
I0206 17:11:06.031324 49646 storage/split_queue.go:97  splitting range=1 [/Min-/Max) at keys [/Table/11/0 /Table/12/0]
I0206 17:11:06.031422 49646 server/node.go:446  connecting to gossip network to verify cluster ID...
I0206 17:11:06.031447 49646 server/node.go:463  node connected via gossip and verified as part of cluster "b2a4b220-9503-419e-acc0-7c0c7f31d6f9"
I0206 17:11:06.031538 49646 server/node.go:298  Started node with [[hdd]=/Users/jaten/cockroachdb.data] engine(s) and attributes []
I0206 17:11:06.032155 49646 server/server.go:262  starting https server at [::]:26257
I0206 17:11:06.032167 49646 storage/stores.go:292  wrote 1 node addresses to persistent storage
I0206 17:11:06.032604 49646 sql/pgwire/server.go:94  starting postgres server at [::]:15432
W0206 17:11:06.033808 49646 rpc/client.go:326  x509: certificate is valid for 127.0.0.1, not ::
I0206 17:11:06.058319 49646 server.go:1934  http: TLS handshake error from [::1]:56449: remote error: bad certificate
W0206 17:11:07.104533 49646 rpc/client.go:326  x509: certificate is valid for 127.0.0.1, not ::
I0206 17:11:07.108192 49646 server.go:1934  http: TLS handshake error from [::1]:56453: remote error: bad certificate
W0206 17:11:08.880371 49646 rpc/client.go:326  x509: certificate is valid for 127.0.0.1, not ::
I0206 17:11:08.885337 49646 server.go:1934  http: TLS handshake error from [::1]:56456: remote error: bad certificate
  C-c C-cI0206 17:11:10.383716 49646 cli/start.go:148  initiating graceful shutdown of server
I0206 17:11:10.383829 49646 stopper.go:273  draining; tasks left:
1      storage/queue.go:334
E0206 17:11:11.033192 49646 storage/queue.go:124  [split] range=1 [/Min-/Max): error: storage/split_queue.go:100: unable to split range=1 [/Min-/Max) at key "/Table/11/0": failed to send RPC: too many errors encountered (1 of 1 total): kv/send.go:282: rpc to Node.Batch: client not ready after 5s
I0206 17:11:11.033302 49646 storage/engine/rocksdb.go:140  closing rocksdb instance at "/Users/jaten/cockroachdb.data"
I0206 17:11:11.033520 49646 cli/start.go:172  server drained and shutdown completed
jaten@Jasons-MacBook-Pro:~/go/src/github.com/cockroachdb/cockroach$
~~~

attaching generated certs
[certs.tar.gz](https://github.com/cockroachdb/cockroach/files/120519/certs.tar.gz)

**Comment by glycerine** [comment]
I tried a variation with
~~~
 ./cockroach cert create-node 127.0.0.1 localhost  10.0.0.165
~~~
but I still get cert errors
~~~
W0206 17:19:58.737409 49673 rpc/client.go:326  x509: certificate is valid for 127.0.0.1, 10.0.0.165, not ::
I0206 17:19:58.741327 49673 server.go:1934  http: TLS handshake error from [::1]:56588: remote error: bad certificate
~~~

**Comment by glycerine** [comment]
I'm hacking around, and crazily enough, adding `::` to the create-node command seems to make it happy.

Wishing for documentation on this topic, so this continues to work... or I learn what will continue to work in the future.  Also `::` seems like it might be a security hole in such a cert setup, yes?

~~~
jaten@Jasons-MacBook-Pro:~/go/src/github.com/cockroachdb/cockroach$ rm -rf certs
jaten@Jasons-MacBook-Pro:~/go/src/github.com/cockroachdb/cockroach$ ./cockroach cert create-node 127.0.0.1 localhost  10.0.0.165 ::
Error: failed to generate node certificate: security/certs.go:58: error loading CA certificate certs/ca.crt and key certs/ca.key: open certs/ca.crt: no such file or directory
Failed running "cert"
jaten@Jasons-MacBook-Pro:~/go/src/github.com/cockroachdb/cockroach$ cockroach cert create-ca
jaten@Jasons-MacBook-Pro:~/go/src/github.com/cockroachdb/cockroach$ ./cockroach cert create-node 127.0.0.1 localhost  10.0.0.165 ::
jaten@Jasons-MacBook-Pro:~/go/src/github.com/cockroachdb/cockroach$ cockroach cert create-client root
jaten@Jasons-MacBook-Pro:~/go/src/github.com/cockroachdb/cockroach$ rm -rf /Users/jaten/cockroachdb.data
jaten@Jasons-MacBook-Pro:~/go/src/github.com/cockroachdb/cockroach$ cockroach start --stores=hdd=/Users/jaten/cockroachdb.data 
I0206 17:30:29.803016 49685 cli/start.go:104  [build]  @  (go1.5.1)
I0206 17:30:29.803626 49685 server/context.go:218  1 storage engine(s) specified
I0206 17:30:29.803653 49685 cli/start.go:125  starting cockroach node
I0206 17:30:29.810075 49685 storage/engine/rocksdb.go:106  opening rocksdb instance at "/Users/jaten/cockroachdb.data"
I0206 17:30:29.816183 49685 server/node.go:320  store store=0:0 ([hdd]=/Users/jaten/cockroachdb.data) not bootstrapped
I0206 17:30:29.821643 49685 storage/replica_command.go:1215  range 1: new leader lease replica {1 1 1} 1970-01-01 00:00:00 +0000 UTC +1454808630.820s
I0206 17:30:29.822277 49685 server/node.go:263  **** cluster ba085228-3bfa-4dd3-9e19-9335d72042d1 has been created
I0206 17:30:29.822314 49685 server/node.go:264  **** add additional nodes by specifying --join=[::]:26257
I0206 17:30:29.824138 49685 server/node.go:333  initialized store store=1:1 ([hdd]=/Users/jaten/cockroachdb.data): {Capacity:367416967168 Available:150487326720 RangeCount:0}
I0206 17:30:29.824175 49685 server/node.go:236  node ID 1 initialized
I0206 17:30:29.824583 49685 gossip/gossip.go:218  setting node descriptor node_id:1 address:<network_field:"tcp" address_field:"[::]:26257" > attrs:<> 
I0206 17:30:29.824758 49685 storage/stores.go:277  read 0 node addresses from persistent storage
I0206 17:30:29.824850 49685 storage/split_queue.go:97  splitting range=1 [/Min-/Max) at keys [/Table/11/0 /Table/12/0]
I0206 17:30:29.824915 49685 server/node.go:446  connecting to gossip network to verify cluster ID...
I0206 17:30:29.825268 49685 storage/stores.go:292  wrote 1 node addresses to persistent storage
I0206 17:30:29.825352 49685 server/node.go:463  node connected via gossip and verified as part of cluster "ba085228-3bfa-4dd3-9e19-9335d72042d1"
I0206 17:30:29.825424 49685 server/node.go:298  Started node with [[hdd]=/Users/jaten/cockroachdb.data] engine(s) and attributes []
I0206 17:30:29.826512 49685 server/server.go:262  starting https server at [::]:26257
I0206 17:30:29.827245 49685 sql/pgwire/server.go:94  starting postgres server at [::]:15432
I0206 17:30:29.842279 49685 storage/replica_command.go:1288  initiating a split of range=1 [/Min-/Max) at key /Table/11
I0206 17:30:29.857708 49685 storage/replica_command.go:1215  range 2: new leader lease replica {1 1 1} 1970-01-01 00:00:00 +0000 UTC +1454808631.073s
I0206 17:30:29.857773 49685 storage/replica_command.go:1288  initiating a split of range=2 [/Table/11-/Max) at key /Table/12
E0206 17:30:34.826571 49685 storage/queue.go:124  [replicate] range=1 [/Min-/Table/11): error: storage/allocator.go:254: no suitable replication target store found, are you running enough nodes?
I0206 17:30:39.826272 49685 server/status/runtime.go:137  runtime stats: 50 goroutines, 3.64MiB active, 0.00cgo/sec, 0.00/0.00 %(u/s)time, 0.00 %gc (1x)
E0206 17:30:39.826873 49685 storage/queue.go:124  [replicate] range=1 [/Min-/Table/11): error: storage/allocator.go:254: no suitable replication target store found, are you running enough nodes?
E0206 17:30:44.828197 49685 storage/queue.go:124  [replicate] range=2 [/Table/11-/Table/12): error: storage/allocator.go:254: no suitable replication target store found, are you running enough nodes?
I0206 17:30:49.826850 49685 server/status/runtime.go:137  runtime stats: 50 goroutines, 2.73MiB active, 37.70cgo/sec, 0.00/0.00 %(u/s)time, 0.00 %gc (1x)
I0206 17:30:49.828308 49685 storage/replica_command.go:1215  range 3: new leader lease replica {1 1 1} 1970-01-01 00:00:00 +0000 UTC +1454808650.828s
E0206 17:30:49.828413 49685 storage/queue.go:124  [replicate] range=3 [/Table/12-/Max): error: storage/allocator.go:254: no suitable replication target store found, are you running enough nodes?
~~~

**Comment by tamird** [comment]
on your machine, what does `hostname` return?

**Comment by glycerine** [comment]
~~~
jaten@Jasons-MacBook-Pro:~/go/src/github.com/cockroachdb/cockroach$ hostname
Jasons-MacBook-Pro.local
jaten@Jasons-MacBook-Pro:~/go/src/github.com/cockroachdb/cockroach$
~~~

**Comment by glycerine** [comment]
What if my host has more than one network card -- how do I tell cockroachdb to bind a specific IP address?

**Comment by tamird** [comment]
You can pass the `--host` argument.

**Comment by glycerine** [comment]
Ah, --host, there it is! Thank you.

**Comment by glycerine** [comment]
So using host to reflect the IP given when creating the certs makes startup go cleanly -- from the TLS cert perspective.  Worth documenting for sure.
~~~
rm -rf certs
cockroach cert create-ca
cockroach cert create-node 10.0.0.165
cockroach cert create-client root
rm -rf /Users/jaten/cockroachdb.data
cockroach start --stores=hdd=/Users/jaten/cockroachdb.data  --host=10.0.0.165
~~~

output:

~~~
jaten@Jasons-MacBook-Pro:~/go/src/github.com/cockroachdb/cockroach$ ./fresh 
I0206 18:23:04.047682 49769 cli/start.go:104  [build]  @  (go1.5.1)
I0206 18:23:04.047808 49769 server/context.go:218  1 storage engine(s) specified
I0206 18:23:04.047818 49769 cli/start.go:125  starting cockroach node
I0206 18:23:04.050188 49769 storage/engine/rocksdb.go:106  opening rocksdb instance at "/Users/jaten/cockroachdb.data"
I0206 18:23:04.052447 49769 server/node.go:320  store store=0:0 ([hdd]=/Users/jaten/cockroachdb.data) not bootstrapped
I0206 18:23:04.053855 49769 storage/replica_command.go:1215  range 1: new leader lease replica {1 1 1} 1970-01-01 00:00:00 +0\
000 UTC +1454811785.054s
I0206 18:23:04.054138 49769 server/node.go:263  **** cluster 4af74af5-89b6-4717-b885-20e214558c06 has been created
I0206 18:23:04.054163 49769 server/node.go:264  **** add additional nodes by specifying --join=10.0.0.165:26257
I0206 18:23:04.054945 49769 server/node.go:333  initialized store store=1:1 ([hdd]=/Users/jaten/cockroachdb.data): {Capacity:\
367416967168 Available:150873034752 RangeCount:0}
I0206 18:23:04.054964 49769 server/node.go:236  node ID 1 initialized
I0206 18:23:04.055084 49769 gossip/gossip.go:218  setting node descriptor node_id:1 address:<network_field:"tcp" address_fiel\
d:"10.0.0.165:26257" > attrs:<>
I0206 18:23:04.073290 49769 storage/stores.go:277  read 0 node addresses from persistent storage
I0206 18:23:04.073304 49769 storage/split_queue.go:97  splitting range=1 [/Min-/Max) at keys [/Table/11/0 /Table/12/0]
I0206 18:23:04.073396 49769 server/node.go:446  connecting to gossip network to verify cluster ID...
I0206 18:23:04.073413 49769 server/node.go:463  node connected via gossip and verified as part of cluster "4af74af5-89b6-4717\
-b885-20e214558c06"
I0206 18:23:04.073471 49769 server/node.go:298  Started node with [[hdd]=/Users/jaten/cockroachdb.data] engine(s) and attribu\
tes []
I0206 18:23:04.073524 49769 server/server.go:262  starting https server at 10.0.0.165:26257
I0206 18:23:04.073618 49769 storage/stores.go:292  wrote 1 node addresses to persistent storage
I0206 18:23:04.073663 49769 sql/pgwire/server.go:94  starting postgres server at 10.0.0.165:15432
I0206 18:23:04.099354 49769 storage/replica_command.go:1288  initiating a split of range=1 [/Min-/Max) at key /Table/11
I0206 18:23:04.104707 49769 storage/replica_command.go:1215  range 2: new leader lease replica {1 1 1} 1970-01-01 00:00:00 +0\
000 UTC +1454811785.304s
I0206 18:23:04.104780 49769 storage/replica_command.go:1288  initiating a split of range=2 [/Table/11-/Max) at key /Table/12
...
~~~

**Comment by jseldess** [comment]
@glycerine, Just updated [Start a Local Cluster](https://www.cockroachlabs.com/docs/start-a-local-cluster.html), and recently added docs for [`cockroach start`](https://www.cockroachlabs.com/docs/start-a-node.html) and [`cockroach cert`](https://www.cockroachlabs.com/docs/create-security-certificates.html). Feedback very welcome (best to open an issue in [docs repo](https://github.com/cockroachdb/docs/issues)).
